Sstrehlk/anomaly segmentation train tests

sstrehlk opened the pull request:
Added precommit and reallife training tests for Anomaly Segmentation models:
- ote_anomaly_segmentation_padim, 
- ote_anomaly_segmentation_stfpm
Task: CVS-81381
Tested on mvtec dataset
NNCF skipped
Evaluation is failed because task tries to compute localization performance while no localization (bbox) annotations are available. In result performance is 0. This is a temporary issue that will be solved once we merge [this PR](https://github.com/openvinotoolkit/training_extensions/pull/993) on the OTE side. The PR adds multi-score performance functionality. Once it's merged, the task will detect if no local annotations are available, and report the image-level performance instead."

build: https://ci-ote.iotg.sclab.intel.com/job/ote/job/ote-test/1578/ 
NNCF not enabled for anomaly yet.

Commit:
ci configs

Commit:
openvino version

Commit:
versions

Commit:
config

Commit:
Added Anomaly real-life training tests

Commit:
Applying comments v.1

Commit:
Applying comments v2

Commit:
Merge branch 'develop' into pf/adding_anomaly_training_tests

sstrehlk committed:
Anomaly Segmentation train tests

sstrehlk committed:
Anomaly Segmentation train tests (extended debug)

sstrehlk committed:
Anomaly Segmentation train tests
Added setting optimization type in BaseAnomalyTask class

sstrehlk committed:
Anomaly Segmentation train tests
Added setting optimization type in BaseAnomalyTask class

sstrehlk committed:
Anomaly Segmentation train tests
Added some debug

sstrehlk committed:
Anomaly Segmentation train tests
Added some debug

sstrehlk committed:
Anomaly Segmentation train tests
Added import of ModelFormat and ModelOptimizationType

sstrehlk committed:
Anomaly Segmentation train tests
Some cleaning

samet-akcay commented on external/anomaly/ote_tests_pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
+python_files = test_ote_anomaly_segmentation_training.py
```
missing new line here.

samet-akcay commented on external/anomaly/tests/expected_metrics/metrics_test_ote_training.yml:
```diff
@@ -0,0 +1,43 @@
+'ACTION-nncf_export_evaluation,model-ote_anomaly_segmentation_stfpm,dataset-mvtec,patience-CONFIG,batch-CONFIG,usecase-reallife':
+        'metrics.accuracy.f-measure':
+                'base': 'nncf_evaluation.metrics.accuracy.f-measure'
+                'max_diff': 0.01
```
new line is missing here.

sstrehlk replied:
added

samet-akcay commented on external/anomaly/tests/conftest.py:
```diff
@@ -0,0 +1,53 @@
+# See the License for the specific language governing permissions
+# and limitations under the License.
+
+from ote_sdk.test_suite.pytest_insertions import *
```
Are the functions imported from `from ote_sdk.test_suite.pytest_insertions` too many? Would be good to avoid wildcard import. It would improve the readability, we would know exactly which functions are imported from there.

sstrehlk replied:
* replaced with functions

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+    return root
+
+def DATASET_PARAMETERS_FIELDS() -> List[str]:
+    return deepcopy(['dataset_path',
+                     ])
```
any more params here? If not, we could maybe fit into a single line such as 

`return deepcopy(['dataset_path'])`

sstrehlk replied:
corrected

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+    if dataset_name not in dataset_definitions:
+        raise ValueError(f'dataset {dataset_name} is absent in dataset_definitions, '
+                         f'dataset_definitions.keys={list(dataset_definitions.keys())}')
+    cur_dataset_definition = dataset_definitions[dataset_name]
```
I'm having trouble understanding what `cur` means in `cur_dataset_definition`? Would it be possible to rename the variable without abbreviations?

sstrehlk replied:
renamed

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+DatasetParameters = namedtuple('DatasetParameters', DATASET_PARAMETERS_FIELDS())
+
+
+def _get_dataset_params_from_dataset_definitions(dataset_definitions, dataset_name):
```
Would be good to annotate function arguments with type hints. Would be easier to follow. For example, is `dataset_definitions` `List[str]` and `dataset_name` `str`?

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+    return params
+
+
+def _create_anomaly_segmentation_dataset_and_labels_schema(dataset_params, dataset_name):
```
Same as above, type hints?

sstrehlk replied:
added type hints

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+
+# TODO:sstrehlk: This function copies with minor change OTETestTrainingAction
+#             from ote_sdk.test_suite.
+#             Try to avoid copying of code.
```
Is there a Jira ticket for this for reference? and for other todos in the code?

sstrehlk replied:
I removed this comment, it was based on another part of code. I'don't know if Jira tickets were created for these todos.

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+    _name = "training"
+
+    def __init__(
+        self, dataset, labels_schema, template_path, patience, batch_size
```
type hints?

sstrehlk replied:
added type hints

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+            raise RuntimeError("Cannot get training performance")
+        return performance_to_score_name_value(training_performance)
+
+    def _run_ote_training(self, data_collector):
```
`data_collector: DataCollector`?

sstrehlk replied:
corrected

samet-akcay commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,420 @@
+                'batch_size': batch_size,
+            }
+
+        def _nncf_graph_params_factory() -> Dict:
```
Looks like almost the same variables are returned as in `params_factories_for_test_actions_fx`. Perhaps we could use something similar `Dict[str,Callable[[], Dict]]` rather than just `Dict`?

sstrehlk replied:
Corrected

samet-akcay reviewed (changes requested):
Thanks for adding the tests @sstrehlk! Overall, looks good, just couple of comments. 

Apart from the reviews I added to the code, I have another comment regarding the formatting. I'm aware that currently tests are not formatted. However, it would be good to stick to a certain style for formatting rather than using multiple styles, which could be seen throughout the code. 

To avoid this, I would run the following from the terminal
```bash
pip install black
black -l 120 external/anomaly/tests/conftest.py
black -l 120 external/anomaly/tests/test_ote_anomaly_segmentation_training.py
```

My last comment is, I'm wondering whether the coverage is taken into account. If so, what percentage is our target?

Thanks again!

Commit:
Merge branch 'develop' into pf/adding_anomaly_training_tests

sstrehlk committed:
Anomaly Segmentation train tests
Corrections after review.

sstrehlk committed:
Anomaly Segmentation train tests

sstrehlk commented:
run ote-test

sstrehlk committed:
Anomaly Segmentation train tests
review corrections

sstrehlk commented:
run ote-test

Commit:
Applying comments v2

Commit:
Added NNCF steps

Commit:
Merge branch 'develop' into pf/adding_anomaly_training_tests

Commit:
Added x-fail ticket

LeonidBeynenson commented on ote_sdk/ote_sdk/test_suite/training_tests_actions.py:
```diff
@@ -259,10 +259,12 @@ def run_export(environment, dataset, task, action_name, expected_optimization_ty
         exported_model.model_format == ModelFormat.OPENVINO
-    ), f"In action '{action_name}': Wrong model format after export"
+    ), f"In action '{action_name}': Wrong model format after export. " \
+       f"exported_model.model_format=exported_model.model_format"
```
```suggestion
       f"{exported_model.model_format=}"
```
-- this is [f-strings with equal sign for debugging](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging)

LeonidBeynenson commented:
run ote_sdk tests

LeonidBeynenson commented on external/anomaly/ote_tests_pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
+python_files = test_ote_anomaly_segmentation_training.py
```
```suggestion
python_files = test_ote_*.py
```

LeonidBeynenson replied:
(You could commit this at the end of developing, -- I understand that this trick increases speed of test jobs)

LeonidBeynenson commented on external/anomaly/tests/test_ote_anomaly_segmentation_training.py:
```diff
@@ -0,0 +1,417 @@
+             test_case_fx, data_collector_fx,
+             cur_test_expected_metrics_callback_fx):
+        if "nncf" in test_parameters["test_stage"]:
+            pytest.xfail("NNCF not yet supported for Anomaly Segmentation")
```
Since this week we do not add XFail mark without a Jira ticket. Please, point a Jira ticket for adding NNCF.

LeonidBeynenson reviewed (changes requested):
Generally the code is absolutely OK, but we should wait until we run reallife training by these tests and check them how they work.

Also ote_sdk tests may be failed due to linters -- the linters should be fixed too
(but, probably, Pavel Finashov will fix some of the linter issues in his branch)

Commit:
Fixed linters

Commit:
Merge branch 'develop' into pf/adding_anomaly_training_tests

Commit:
Fix pylint

sstrehlk committed:
Anomaly Segmentation train tests
review corrections

sstrehlk committed:
Merge remote-tracking branch 'origin/pf/adding_anomaly_training_tests' into sstrehlk/anomaly_segmentation_train_tests

# Conflicts:
#	external/anomaly/ote_tests_pytest.ini
#	external/anomaly/tests/conftest.py
#	external/anomaly/tests/expected_metrics/metrics_test_ote_training.yml

sstrehlk committed:
Anomaly Segmentation train tests
Correction in fixtures

sstrehlk committed:
Anomaly Segmentation train tests
Corrected TaskType argument for OteMvtecDataset class for anomaly segmentation tests.
Blocked anomaly classification tests to limit debug scope
Temporary blocked real life dataset.

sstrehlk commented:
run ote-test

sstrehlk commented:
run ote-sdk tests

sstrehlk committed:
Anomaly Segmentation train tests
Enabled real life dataset.
Disabled xfail for nncf.

sstrehlk committed:
Anomaly Segmentation train tests
Blocked stfpm model - for debug purposes.

sstrehlk committed:
Anomaly Segmentation train tests
enabled api_tests and inference tests

sstrehlk committed:
Merge branch 'develop' into sstrehlk/anomaly_segmentation_train_tests

sstrehlk committed:
Merged changes from develop branch

sstrehlk commented:
Anomaly Classification and Anomaly Segmentation were split into separate  branches and PRs, so I'm closing this PR.
